main/streams/userspace.c

```diff
@@ -836,205 +812,246 @@ static void statbuf_from_array(zval *array, php_stream_statbuf *ssb)
+
+	if (UNEXPECTED(call_result == FAILURE)) {
+		php_error_docref(NULL, E_WARNING,
+				"%s::" USERSTREAM_EOF " is not implemented! Assuming EOF",
+				ZSTR_VAL(us->wrapper->ce->name));
+		// TODO: Returning PHP_STREAM_OPTION_RETURN_NOTIMPL means the stream_set_option() returns true...
+		return PHP_STREAM_OPTION_RETURN_ERR;
+	}
+	if (UNEXPECTED(Z_ISUNDEF(retval))) {
+		return PHP_STREAM_OPTION_RETURN_ERR;
+	}
+	if (EXPECTED(Z_TYPE(retval) == IS_FALSE || Z_TYPE(retval) == IS_TRUE)) {
+		// TODO This seems wrong?
+		return Z_TYPE(retval) == IS_TRUE ? PHP_STREAM_OPTION_RETURN_ERR : PHP_STREAM_OPTION_RETURN_OK;
+	} else {
+		php_error_docref(NULL, E_WARNING,
+			"%s::" USERSTREAM_EOF " value must be of type bool, %s given",
+				ZSTR_VAL(us->wrapper->ce->name), zend_zval_value_name(&retval));
 		zval_ptr_dtor(&retval);
-		zval_ptr_dtor(&func_name);
-		break;
+		return PHP_STREAM_OPTION_RETURN_ERR;
+	}
+}
 
-	case PHP_STREAM_OPTION_LOCKING:
-		ZVAL_LONG(&args[0], 0);
+static int user_stream_set_locking(const php_userstream_data_t *us, int value)
+{
+	zval retval;
+	zval zlock;
+	zend_long lock = 0;
 
-		if (value & LOCK_NB) {
-			Z_LVAL_P(&args[0]) |= PHP_LOCK_NB;
-		}
-		switch(value & ~LOCK_NB) {
+	if (value & LOCK_NB) {
+		lock |= PHP_LOCK_NB;
+	}
+	switch (value & ~LOCK_NB) {
 		case LOCK_SH:
-			Z_LVAL_P(&args[0]) |= PHP_LOCK_SH;
+			lock |= PHP_LOCK_SH;
 			break;
 		case LOCK_EX:
-			Z_LVAL_P(&args[0]) |= PHP_LOCK_EX;
+			lock |= PHP_LOCK_EX;
 			break;
 		case LOCK_UN:
-			Z_LVAL_P(&args[0]) |= PHP_LOCK_UN;
+			lock |= PHP_LOCK_UN;
 			break;
-		}
+		default:
+			// TODO: Warn on invalid option value?
+			;
+	}
+	ZVAL_LONG(&zlock, lock);
 
-		/* TODO wouldblock */
-		ZVAL_STRINGL(&func_name, USERSTREAM_LOCK, sizeof(USERSTREAM_LOCK)-1);
-
-		call_result = call_method_if_exists(&us->object, &func_name, &retval, 1, args);
-
-		if (call_result == SUCCESS && (Z_TYPE(retval) == IS_FALSE || Z_TYPE(retval) == IS_TRUE)) {
-			ret = (Z_TYPE(retval) == IS_FALSE);
-		} else if (call_result == FAILURE) {
-			if (value == 0) {
-			   	/* lock support test (TODO: more check) */
-				ret = PHP_STREAM_OPTION_RETURN_OK;
-			} else {
-				php_error_docref(NULL, E_WARNING, "%s::" USERSTREAM_LOCK " is not implemented!",
-								 ZSTR_VAL(us->wrapper->ce->name));
-				ret = PHP_STREAM_OPTION_RETURN_ERR;
-			}
-		}
+	/* TODO wouldblock */
+	zend_string *func_name = ZSTR_INIT_LITERAL(USERSTREAM_LOCK, false);
+	zend_result call_result = zend_call_method_if_exists(Z_OBJ(us->object), func_name, &retval, 1, &zlock);
+	zend_string_release_ex(func_name, false);
 
+	if (UNEXPECTED(call_result == FAILURE)) {
+		if (value == 0) {
+			/* lock support test (TODO: more check) */
+			return PHP_STREAM_OPTION_RETURN_OK;
+		}
+		php_error_docref(NULL, E_WARNING,
+				"%s::" USERSTREAM_LOCK " is not implemented!",
+				ZSTR_VAL(us->wrapper->ce->name));
+		// TODO: Returning PHP_STREAM_OPTION_RETURN_NOTIMPL means the stream_set_option() returns true...
```

Comment:
Not sure that's the case for locking. Basically `php_stream_lock` is success only if 0 by the look of it. So both are the same. But as I said above NOTIMPL is usually just for missing functionality and it doesn't usually trigger error.